client/src/r_draw.cpp

```diff
@@ -454,44 +472,51 @@ void R_CopyTranslationRGB (int fromplayer, int toplayer)
 
 // [RH] Create a player's translation table based on
 //		a given mid-range color.
-void R_BuildPlayerTranslation(int player, argb_t dest_color)
+void R_BuildPlayerTranslation(int player, argb_t dest_color, int colorpreset)
 {
-	const palette_t* pal = V_GetDefaultPalette();
-	byte* table = &translationtables[player * 256];
-
-	fahsv_t hsv_temp = V_RGBtoHSV(dest_color);
-	float h = hsv_temp.geth(), s = hsv_temp.gets(), v = hsv_temp.getv();
-
-	s -= 0.23f;
-	if (s < 0.0f)
-		s = 0.0f;
-	float sdelta = 0.014375f;
-
-	v += 0.1f;
-	if (v > 1.0f)
-		v = 1.0f;
-	float vdelta = -0.05882f;
-
-	for (int i = 0x70; i < 0x80; i++)
+	if (colorpreset > -1 && colorpreset < 6)
```

Comment:
Can we use an enum value to specify that a preset is invalid (eg, the player is not using a preset)? Often times, the last element in the enum is used to represent the invalid option. If `colorpreset` is changed to be an unsigned int, you can simply have the check `if (colorpreset < NO_COLORPRESET)`. That would remove two magic numbers from this line of code.

Reply:
Additionally, the logic to check if this is a team game (which would override the color preset) could go into this check. That would simplify all of the calls to `R_BuildPlayerTranslation()` because they wouldn't have to use the `G_IsTeamColor()` function to provide the value for `colorpreset`.

Reply:
A distinction here is that there are actually two types of colour presets here. Currently, there are just colour presets, but the idea here is to convert _some_ of them to use the Vanilla palette indexing rather than RGB values. https://github.com/Acts19quiz/odamex_Acts19quiz/blob/808bc8738de81ee247e6e5a8229e2f1e8e2f42a1/client/src/d_netinfo.cpp#L123-L148 COLOR_GREEN (0) through COLOR_ORANGE (5) are Vanilla presets, while everything else are non-Vanilla presets that, while being colour presets, we don't using Vanilla indexing. COLOR_PURPLE, for example, is a preset we still want RGB translation on. As well as COLOR_CUSTOM, obviously.

Given what you're suggesting in the first post, would Odamex still be able to make this distinction (i.e., it's in the 0-5 range)?

Reply:
Regarding the second post: I just moved the r-forceteamcolor and r-forceenemycolor checks to a single G_IsTeamColor call in R_BuildPlayerTranslation. This introduced a bug where activating r-forceteamcolor (in a coop game) removes the Vanilla indexing for the player using it as well as the other players. R_BuildPlayerTranslation says, 'Oh, Player A has r-forceteamcolor on! Better stop *all* Vanilla translations!' without discriminating what sprites to stop the Vanilla translation for (Player B and C, and only from A's POV). If Player A has r-forceteamcolor on in COOP but is using Vanilla "red", we want to turn off Vanilla translation for B and C from A's POV, but still have A use Vanilla translation. The following calls to R_BuildPlayerTranslation need to individually collect the relevant player's r- values:

https://github.com/Acts19quiz/odamex_Acts19quiz/blob/808bc8738de81ee247e6e5a8229e2f1e8e2f42a1/client/src/cl_main.cpp#L239
https://github.com/Acts19quiz/odamex_Acts19quiz/blob/808bc8738de81ee247e6e5a8229e2f1e8e2f42a1/client/src/cl_main.cpp#L1443

r_draw.cpp can't check for r- because it just can't know. What I *could* do, since only two calls to R_BuildPlayerTranslation need to check for r- values, is have those two lines make additional checks. Something like this (from cl_main.cpp#L1443):

```
	else
	{
		R_BuildPlayerTranslation(player.id, CL_GetPlayerColor(&player), G_IsTeamColor(player.userinfo.colorpreset, r_forceteamcolor, r_forceenemycolor));
	}
```

to

```
	else
	{
		if ((r_forceteamcolor && G_IsCoopGame()) || (r_forceenemycolor && G_IsFFAGame()))
			R_BuildPlayerTranslation(player.id, CL_GetPlayerColor(&player), NULLVALUE);
		else
			R_BuildPlayerTranslation(player.id, CL_GetPlayerColor(&player), player.userinfo.colorpreset);
	}
```

Then r_draw.cpp can check for CTF or TEAMDM since those are the two circumstances where we actually want to turn Vanilla translations off all the time. At the expense of two extra if/else checks, all R_BuildPlayerTranslation calls are simplified and need to only pass colorpreset, not three different values to G_IsTeamColor.

Reply:
> Given what you're suggesting in the first post, would Odamex still be able to make this distinction (i.e., it's in the 0-5 range)?

I think you can accomplish this by sub-dividing the enum a bit
```enum colorpreset_t	// Acts 19 quiz the order must match m_menu.cpp.
{
	COLOR_GREEN,
	COLOR_INDIGO,
	COLOR_BROWN,
	COLOR_RED,
	COLOR_BLUE,
	COLOR_ORANGE,
	INDEXED_COLOR_PRESET_RANGE,
	
	COLOR_GOLD,
	COLOR_JUNGLEGREEN,
	COLOR_PURPLE,
	COLOR_WHITE,
	COLOR_BLACK,
	COLOR_CUSTOM,

	ARGB_COLOR_PRESET_RANGE
}```

So you can use these conditionals now:
`if (colorpreset < INDEXED_COLOR_PRESET_RANGE) ...`
or
`else if (colorpreset < ARGB_COLOR_PRESET_RANGE)...`

Reply:
It looks like the parameters that are passed to `R_BuildPlayerTranslation()` can all be obtained from a `player_t` object. I propose changing it to have a single `player_t` parameter and extracting the player ID, color preset ID, and ARGB translation value inside `R_BuildPlayerTranslation()`. Then `R_BuildPlayerTranslation()` can own all of the logic for determining how to translate the colors for that player.

`R_BuildPlayerTranslation` should be able to look at the color preset, the ARGB translation, the game mode and the force team/enemy color CVARs and decide how to translate the color for the given player, right?

Reply:
1) I tired this change out, and it's... problematic. I inserted the sub-divider into d_netinf.h, but this causes a crash when messing around in PLAYER SETUP. As my note there observes, the list must match `static const char *colorpresets` in m_menu.cpp (and vice versa). So I have to put the sub-divider in that list of options, as well. And doing that makes it want to have it as a selectible option in PLAYER SETUP, which we obviously don't want (and if I don't do that, `static void M_ChangeColorPreset (int choice)` goes haywire--and crashes). Maybe I did it wrong, but whereas I can truncate the list of options to ignore the last one (NUMCOLOR), this is right in the middle of options.

With enum usage akin to what you're suggesting, I was able to eliminate the `colorpreset > -1` check, but with an end-of-enum-list definition being the total (12) when we want 6, the best I can think of is something like `#define MAXVANILLA 6` in r_draw.cpp. That's probably not quite what you want, but it's the closest thing I can come up with on the spot.

2) If I understand correctly, what you're suggesting is effectively "Proposal D", transforming R_BuildPlayerTranslation from receiving `int player` to `player_t player`.